### PR TITLE
Redistribute command distributions

### LINKED ISSUE #11914: Create a CommandRedistributor
We need to be able to redistribute pending command distributions. This will work in a similar way as the [DeploymentRedistributor](https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/distribute/DeploymentRedistributor.java).

**Always**
- Create `CommandRedistributor`
- Remove `@Ignore` in `MultiPartitionDeploymentLifecycleTest.shouldRejectCompleteDeploymentDistributionWhenAlreadyCompleted`

**Option 1**
- Deprecate `DeploymentRedistributor`, we need to keep it to stay backwards compatible
- See if we can disable the `DeploymentRedistributor` from running if there are no pending deployments

**Option 2**
- Migrate pending deployment distributions to generalised distribution

### PR DESCRIPTION
**Pull request by korthout**
## Description

Introduces a new CommandRedistributor class that retries sending pending command distributions to other partitions. An exponential backoff is used, as it is almost a 1:1 copy of the DeploymentRedistributor.

In addition, the redistribution has been made ready to be scheduled as async tasks. For now, it does not yet run async, as this will be added using a feature flag in the future (#12308).

Not all functionalities of this can be tested yet, but they will be tested when we switch over Deployment Distribution to the new Command Distribution (#11962).

## Related issues

closes #11914 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/state/distribution/DbDistributionState.java [hunk 1]
```diff
@@ -128,4 +133,35 @@ public CommandDistributionRecord getCommandDistributionRecord(
...
+  public void foreachPendingDistribution(final PendingDistributionVisitor visitor) {
+    final var lastDistributionKey = new MutableLong(0);
+    final var lastPendingDistribution = new MutableReference<CommandDistributionRecord>();
+
+    pendingDistributionColumnFamily.forEach(
+        (compositeKey, nil) -> {
+          final var distributionKey = compositeKey.first().inner().getValue();
+          final var partitionId = compositeKey.second().getValue();
+
+          if (lastDistributionKey.value != distributionKey) {
```
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/state/distribution/DbDistributionState.java:147]: ❓ Why do we have to keep track of the last distribution and perform this check?

File: engine/src/test/java/io/camunda/zeebe/engine/state/distribution/DistributionStateTest.java [hunk 2]
```diff
@@ -165,10 +170,107 @@ public void shouldFailToAddPendingDistributionIfNoCommandDistributionExists() {
...
+    final var distributions = new Long2ObjectHashMap<CommandDistributionRecord>();
+    final int partitionId2 = 2;
+    final int partitionId3 = 3;
+    for (int distributionKey = 1; distributionKey <= 5; distributionKey++) {
+      final var pendingDistribution = createCommandDistributionRecord();
+      distributions.put(distributionKey, pendingDistribution);
+      distributionState.addCommandDistribution(distributionKey, pendingDistribution);
+
+      distributionState.addPendingDistribution(distributionKey, partitionId2);
+      distributionState.addPendingDistribution(distributionKey, partitionId3);
```
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/state/distribution/DistributionStateTest.java:218]: 🔧 We have to do these calls in all these test cases. Might be nice to extract it to a separate method.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/distribution/CommandRedistributor.java [hunk 3]
```diff
@@ -9,35 +9,66 @@
...
+ * CommandDistributionRecord}s to other partitions. This is needed because the communication between
+ * partitions is unreliable.
+ *
+ * <p>A simple exponential backoff is used for retrying these pending distributions. This
+ * exponential backoff is statically configured to start of at 10 seconds {@link
+ * #COMMAND_REDISTRIBUTION_INTERVAL} until it reaches a maximum of 5 minutes {@link
+ * #RETRY_MAX_BACKOFF_DURATION}, doubling every time. This backoff is tracked for each pending
+ * distribution individually.
+ */
+public final class CommandRedistributor implements StreamProcessorLifecycleAware {
```
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/distribution/CommandRedistributor.java:34]: 👍 Thanks for adding all the javadoc to the redistributor. Makes it a lot more easy to understand!

**Review by remcowesterhoud — APPROVED** [review]
Thanks @korthout, very clean! I have some minor comments.

❓ In the issue I wrote:

> See if we can disable the `DeploymentRedistributor` from running if there are no pending deployments

Do you think it makes sense to do this, maybe as a separate PR. I'm not sure how much it matters if we keep this redistributor going if there's no data in it anymore.

File: engine/src/main/java/io/camunda/zeebe/engine/state/distribution/DbDistributionState.java (hunk 1 above)
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/state/distribution/DbDistributionState.java:147]: It's for performance: we may encounter the same pending distribution for separate partitions. There's no need to look up the command distribution record again in that case, but we can reuse it.

❓ Should I document this somewhere? Perhaps in a comment at that check?

**Comment by korthout** [comment]
@remcowesterhoud What do you think of this?

**Review by remcowesterhoud — APPROVED** [review]
All looks good to me. Thanks @korthout!

bors merge

**Pull request merged**